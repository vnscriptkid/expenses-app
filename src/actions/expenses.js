import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD EXPENSE
export const addExpense = (expense) => (
    {
        type: 'ADD_EXPENSE',
        expense
    }
)

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        })
    }
}

// REMOVE EXPENSE
export const removeExpense = ( id ) => ({
    type: 'REMOVE_EXPENSE',
    id: id
})

export const startRemoveExpense = ( {id} = {} ) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
        .then(() => {
            dispatch(removeExpense(id));
        })
        // .catch((err) => {
        //     console.log(err);
        // })
    }
}

// EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

// SET EXPENSES
export const setExpenses = (expenses) =>({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return (dispatch) => {
        // retrive data from db 
        return database.ref('expenses').once('value').then( (snap) => {
            const expensesObject = snap.val();
            // convert object to array
            const expensesArray = Object.keys(expensesObject).map((key) => ({...expensesObject[key], id: key}));
            // dispatch setExpenses with that array
            dispatch(setExpenses(expensesArray))
        })
    }
}