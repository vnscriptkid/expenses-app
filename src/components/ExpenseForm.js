import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            error: '',
            date: props.expense ? moment(props.expense.createdAt) : moment(),
            focused: false
        }
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onFormSubmit = this.onFormSubmit.bind(this);
    }

    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }))
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange(e) {
        const amount = e.target.value;
        this.setState(() => ({ amount }));
    }

    onFormSubmit(e) {
        e.preventDefault();
        if (this.state.description && parseFloat(this.state.amount) > 0) {
            this.props.onFormSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                note: this.state.note,
                createdAt: this.state.date.valueOf()               
            })
            this.setState(() => ({
                error: '',
                description: '',
                amount: '',
                note: ''
            }))
        } else {
            this.setState(() => ({ error: 'You need to provide proper values for description and amount' }))
        }
    }

    render() {
        return (
            <form onSubmit={this.onFormSubmit}>
                <input
                    type="text"
                    value={this.state.description}
                    placeholder="Description"
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="number"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <textarea
                    placeholder="Note something"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <SingleDatePicker
                    date={this.state.date} // momentPropTypes.momentObj or null
                    onDateChange={date => this.setState({ date })} // PropTypes.func.isRequired
                    focused={this.state.focused} // PropTypes.bool
                    onFocusChange={({ focused }) => this.setState({ focused })} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <br />
                <button>Add Expense</button>
                {this.state.error && <h3 style={{ color: 'red' }}>{this.state.error}</h3>}
            </form>
        )
    }
}
