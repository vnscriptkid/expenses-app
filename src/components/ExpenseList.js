import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import {connect} from 'react-redux';
import ExpensesSelect from '../selectors/expenses';

export const ExpenseList = (props) => {      
    return (
        <div>
            <h3>Expense List</h3>
            { 
                props.expenses.length > 0 ? (
                    <p>There are {props.expenses.length} item(s) in the expense list</p>,
                    props.expenses.map((expense, index) => (                
                        <ExpenseListItem 
                            id={expense.id} 
                            key={expense.id} 
                            count={index+1} 
                            {...expense}
                        />
                    ))
                ) : (
                    <h4>There is no a single one Expense</h4>
                )
            }                                   
        </div>
    );
} 

const mapStateToProps = (state) => {    
    return {
        expenses: ExpensesSelect(state.expenses, state.filters) 
    }
}

// export default ExpenseList;
export default connect(mapStateToProps)(ExpenseList);