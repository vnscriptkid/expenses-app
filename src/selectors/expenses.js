
// Get the expenses then filter following conditions passed in, finally sort the outcome with sortBy
export default (expenses, { text, sortBy, startDate, endDate }) => {    
    return expenses.filter((expense) => {
        // get all the valid expenses that fit the date and match the text
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt > startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt < endDate;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        // finally sort the filtered expenses from previous step and sort depend on value in sortBy
        if (sortBy === 'date')
            return a.createdAt - b.createdAt;
        else if (sortBy === 'amount')
            return a.amount - b.amount;
    })
}