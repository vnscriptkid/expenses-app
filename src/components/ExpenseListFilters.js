import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';
// import 'react-dates/initialize';
// import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            focusedInput: null
        }
        this.onTextChange = this.onTextChange.bind(this);
        this.onSortChange = this.onSortChange.bind(this);
        this.onDatesChange = this.onDatesChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
    }

    onTextChange(e) {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange(e) {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }

    onDatesChange({startDate, endDate}) {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange(focusedInput) {
        this.setState(() => ({ focusedInput }))
    }

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />

                <select onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                
                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
        
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

// export default ExpenseListFilters;
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);