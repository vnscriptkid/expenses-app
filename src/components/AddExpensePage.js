import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

const AddExpensePage = (props) => (
    <div>
        <h2>ADD EXPENSE PAGE</h2>
        <ExpenseForm onFormSubmit={(expense) => {
            props.dispatch(addExpense(expense));
            props.history.push('/');
            console.log(props.history);
            console.log(expense);
        }}/>
    </div>
)

export default connect()(AddExpensePage);
// export default AddExpensePage;
