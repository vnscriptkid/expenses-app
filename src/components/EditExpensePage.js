import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
        this.onEditSubmit = this.onEditSubmit.bind(this);
        this.onRemoveClick = this.onRemoveClick.bind(this);
    }

    onEditSubmit( expense ) {
        this.props.editExpense(
            this.props.expense.id,
            {
                description: expense.description,
                note: expense.note,
                amount: expense.amount,
                createdAt: expense.createdAt
            }
        )
        this.props.history.push('/');
    }

    onRemoveClick() {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <button onClick={this.onRemoveClick}>REMOVE EXPENSE</button>
                <ExpenseForm
                    functionButton="EDIT EXPENSE"
                    expense={this.props.expense}
                    onFormSubmit={this.onEditSubmit}
                />
            </div>
        )
    }
}

// const EditExpensePage = (props) => {
//     return (
//         <div>      
//             <button onClick={() => {
//                 props.dispatch(removeExpense({ id: props.expense.id }));
//                 props.history.push('/');
//             }}>REMOVE EXPENSE</button>     
//             <ExpenseForm 
//                 expense = {props.expense}
//                 onFormSubmit = {(expense) => {
//                     props.dispatch(editExpense(
//                         props.expense.id,
//                         {
//                             description: expense.description,
//                             note: expense.note,
//                             amount: expense.amount,
//                             createdAt: expense.createdAt
//                         }
//                     ))
//                     props.history.push('/');
//                 }}
//             />
//         </div>
//     )
// }


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, updates) => dispatch(editExpense(id, updates)),
    removeExpense: ( idObject ) => dispatch(startRemoveExpense( idObject ))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
// export default EditExpensePage;