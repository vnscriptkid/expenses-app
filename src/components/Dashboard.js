import React from 'react';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseList from './ExpenseList';
import { connect } from 'react-redux';

export const Dashboard = (props) => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default Dashboard;