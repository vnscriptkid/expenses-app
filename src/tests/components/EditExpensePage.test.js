import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';
import ExpenseForm from '../../components/ExpenseForm';

let editExpenseSpy, removeExpenseSpy, historySpy, wrapper;

beforeEach(() => {
    editExpenseSpy = jest.fn();
    removeExpenseSpy = jest.fn();
    historySpy = { push: jest.fn() };
    wrapper = shallow( <EditExpensePage 
        editExpense={editExpenseSpy}
        removeExpense={removeExpenseSpy}
        history={historySpy}
        expense={expenses[0]}
        /> 
    );
}) 

test('should render Edit Expense Page correctly', () => {
    expect(wrapper).toMatchSnapshot();
})

test('should handle editExpense', () => {
    wrapper.find( ExpenseForm ).prop('onFormSubmit')(expenses[0]);
    expect(editExpenseSpy).toHaveBeenCalledWith(
        expenses[0].id,
        {
            description: expenses[0].description,
            note: expenses[0].note,
            amount: expenses[0].amount,
            createdAt: expenses[0].createdAt
        }
    );
    expect(historySpy.push).toHaveBeenCalledWith('/');
})

test('should handle removeExpense', () => {
    wrapper.find('button').simulate('click');
    expect(historySpy.push).toHaveBeenCalledWith('/');
    expect(removeExpenseSpy).toHaveBeenCalledWith({
        id: expenses[0].id
    })
})



