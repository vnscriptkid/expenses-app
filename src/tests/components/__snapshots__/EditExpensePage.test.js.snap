// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit Expense Page correctly 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    REMOVE EXPENSE
  </button>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": "1970-01-01T00:00:00.000Z",
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    functionButton="EDIT EXPENSE"
    onFormSubmit={[Function]}
  />
</div>
`;
